experimental/LieAlgebras/src/LinearLieAlgebra.jl

```diff
@@ -317,37 +321,191 @@ function special_linear_lie_algebra(R::Field, n::Int)
+  for i in 1:n, j in 1:n
+    x = zero_matrix(R, n, n)
+    x[i, j] = 1
+    eqs[(i-1) * n + j, :] = _vec(x + invform * transpose(x) * form)
+  end
+  ker = kernel(eqs)
+  rref!(ker) # we cannot assume anything about the kernel, but want to have a consistent output
+  dim = nrows(ker)
+  basis = [zero_matrix(R, n, n) for _ in 1:dim]
+  for i in 1:n
+    for k in 1:dim
+      basis[k][i, 1:n] = ker[k, (i-1) * n .+ (1:n)]
+    end
+  end
+  return basis
+end
+
 @doc raw"""
-    special_orthogonal_lie_algebra(R::Field, n::Int) -> LinearLieAlgebra{elem_type(R)}
+    special_orthogonal_lie_algebra(R::Field, n::Int, gram::MatElem) -> LinearLieAlgebra{elem_type(R)}
+    special_orthogonal_lie_algebra(R::Field, n::Int, gram::Matrix) -> LinearLieAlgebra{elem_type(R)}
 
 Return the special orthogonal Lie algebra $\mathfrak{so}_n(R)$.
 
+Given a non-degenerate symmetric bilinear form $f$ via its Gram matrix `gram`,
+$\mathfrak{so}_n(R)$ is the Lie algebra of all $n \times n$ matrices $x$ over the field `R`
+such that $f(xv, w) = -f(v, xw)$ for all $v, w \in R^n$.
+
+If `gram` is not provided, for $n = 2k$ the form defined by $\begin{matrix} 0 & I_k \\ -I_k & 0 \end{matrix}$
+is used, and for $n = 2k + 1$ the form defined by $\begin{matrix} 1 & 0 & 0 \\ 0 & 0 I_k \\ 0 & I_k & 0 \end{matrix}$.
+
 # Examples
 ```jldoctest
-julia> L = special_orthogonal_lie_algebra(QQ, 3)
+julia> L1 = special_orthogonal_lie_algebra(QQ, 4)
+Special orthogonal Lie algebra of degree 4
+  of dimension 6
+over rational field
+
+julia> basis(L1)
+6-element Vector{LinearLieAlgebraElem{QQFieldElem}}:
+ x_1
+ x_2
+ x_3
+ x_4
+ x_5
+ x_6
+
+julia> matrix_repr_basis(L1)
+6-element Vector{QQMatrix}:
+ [1 0 0 0; 0 0 0 0; 0 0 -1 0; 0 0 0 0]
+ [0 1 0 0; 0 0 0 0; 0 0 0 0; 0 0 -1 0]
+ [0 0 0 1; 0 0 -1 0; 0 0 0 0; 0 0 0 0]
+ [0 0 0 0; 1 0 0 0; 0 0 0 -1; 0 0 0 0]
+ [0 0 0 0; 0 1 0 0; 0 0 0 0; 0 0 0 -1]
+ [0 0 0 0; 0 0 0 0; 0 1 0 0; -1 0 0 0]
+
+julia> L2 = special_orthogonal_lie_algebra(QQ, 3, identity_matrix(QQ, 3))
 Special orthogonal Lie algebra of degree 3
   of dimension 3
 over rational field
 
-julia> basis(L)
+julia> basis(L2)
 3-element Vector{LinearLieAlgebraElem{QQFieldElem}}:
- x_1_2
- x_1_3
- x_2_3
+ x_1
+ x_2
+ x_3
 
-julia> matrix_repr_basis(L)
+julia> matrix_repr_basis(L2)
 3-element Vector{QQMatrix}:
  [0 1 0; -1 0 0; 0 0 0]
  [0 0 1; 0 0 0; -1 0 0]
  [0 0 0; 0 0 1; 0 -1 0]
 ```
 """
+special_orthogonal_lie_algebra
+
+function special_orthogonal_lie_algebra(R::Field, n::Int, gram::MatElem)
+  form = map_entries(R, gram)
+  @req size(form) == (n, n) "Invalid matrix dimensions"
+  @req is_symmetric(form) "Bilinear form must be symmetric"
+  @req is_invertible(form) "Bilinear form must be non-degenerate"
+  basis = _lie_algebra_basis_from_form(R, n, form)
+  dim = length(basis)
+  @assert characteristic(R) != 0 || dim == div(n^2 - n, 2)
+  s = ["x_$(i)" for i in 1:dim]
+  L = lie_algebra(R, n, basis, s; check=false)
+  set_attribute!(L, :type => :special_orthogonal, :form => form)
```

Comment:
Maybe also add a documented form to query the form? Of course that can also come in a future PR.

Reply:
Comes in a future PR

Reply:
@fingolfin do you have a name idea for this function? `form` can be ambiguous with e.g. the killing form

Reply:
`defining_form`?

Reply:
`invariant_form` (if it is invariant :))?